lib/Arch.cpp

```diff
@@ -0,0 +1,365 @@
+// register allocation.
+remill::Register *TryRegisterAllocate(
+    llvm::Type &type, std::vector<bool> &reserved,
+    const std::vector<RegisterConstraint> &register_constraints) {
+  SizeConstraint size_constraint;
+  TypeConstraint type_constraint;
+
+  if (type.isIntegerTy()) {
+    type_constraint = kTypeInt;
+    auto derived = llvm::cast<llvm::IntegerType>(type);
+    unsigned int width = derived.getBitWidth();
+    if (width == 64) {
+      size_constraint = kMinBit64;
+    } else {
+      // TODO(aty): I know that this is wrong but for now its fine
+      size_constraint = kMinBit32;
+    }
+  } else if (type.isFloatTy()) {
+    type_constraint = kTypeFloat;
+    // We automatically know it is 32-bit IEEE floating point type
+    size_constraint = kMinBit32;
+  } else if (type.isDoubleTy()) {
+    type_constraint = kTypeFloat;
+    // We automatically know it is 64-bit IEEE floating point type
+    size_constraint = kMinBit64;
+  } else if (type.isPointerTy()) {
+    type_constraint = kTypeIntegral;
+    size_constraint = kMinBit64;
+  }
+  // TODO(aty): Handle other types
+
+  for (size_t i = 0; i < register_constraints.size(); i++) {
+    if (reserved[i]) continue;
+
+    const RegisterConstraint &constraint = register_constraints[i];
+    // Iterate through the different sizes of a single register to find the
+    // smallest possible match
+    for (auto const &variant : constraint.variants) {
+      if (size_constraint & variant.size_constraint &&
+          type_constraint & variant.type_constraint) {
+        reserved[i] = true;
+        remill::Register *reg =
+            new remill::Register(variant.register_name, 0, 0, 0, &type);
+        return reg;
+      }
+    }
+  }
+  return nullptr;
+}
+
+// For each element of the struct, try to allocate it to a register, if all of
+// them can be allocated, then return that allocation. Otherwise return a
+// nullptr.
+std::unique_ptr<std::vector<anvill::ValueDecl>> TryReturnThroughRegisters(
+    const llvm::StructType &st,
+    const std::vector<RegisterConstraint> &constraints) {
+  auto ret = std::make_unique<std::vector<anvill::ValueDecl>>();
+  std::vector<bool> reserved(constraints.size(), false);
+  for (unsigned i = 0; i < st.getNumElements(); i++) {
+    anvill::ValueDecl value_decl = {};
+    auto reg =
+        TryRegisterAllocate(*st.getElementType(i), reserved, constraints);
+    if (reg) {
+      value_decl.reg = reg;
+      value_decl.type = st.getElementType(i);
+    } else {
+      // The struct cannot be split over registers
+      return nullptr;
+    }
+    ret->push_back(value_decl);
+  }
+
+  return ret;
+}
+
+std::vector<anvill::ValueDecl> X86_64_SysV::BindReturnValues(
+    const llvm::Function &function) {
+  std::vector<anvill::ValueDecl> return_value_declarations;
+
+  for (auto &block : function) {
+    for (auto &inst : block) {
+      // TODO(aty): There has to be a better way of doing this, right now I am
```

Comment:
I think this would all be simpler if you used `function.getReturnType()`.

Reply:
Fixed 327d728